Semantic search(ncbi file)

Anwesh1 opened the pull request:
The pubmed-semantic.py can be run with the .env from semantic-search as an example of the UID and text tags, just to see that we can't really grab the Title and Abstract separately.
The pubmed-semantic(updated).py is another file that's essentially taking in a list of around 2635 PMID's however there's an error thrown by the uid_to_docs() that's apparently a validation error. This is a little weird because when I use a subset of about 10 entries then it works fine, but I think one of the PMID's doesn't have a title and that's causing an error.

Anwesh1 committed:
UID and text from list of PMID's

If you run it right now, you can see a .json file created that has UID and text tags. If we want the title and abstract separate, I'm not sure how we can do that.

Anwesh1 committed:
Add files via upload

JohnGiorgi commented on pubmed-semantic(updated).py:
```diff
@@ -0,0 +1,286 @@
+    store = []
+    #First loop to grab all the data and write it to .json files, each file only has 10,000 records
+    for i in range(0, total, 10000):
+        #changes the retstart value by 10,000 every iteration so we don't print the same records over again
+        this_fetch = fetch_url+"&retstart="+str(i)
+        #URL that can be used in browser for data viewing
+        print("URL "+str(count)+": "+this_fetch)
+        count += 1
+        fetch_r = requests.post(this_fetch)
+        #Writing all the data to .json files (1 file per 10,000 records)
+        f = open('pubmed_batch_'+str(i)+'_to_'+str(i+9999)+".json", 'w')
+        f.write(fetch_r.text)
+        f.close()
+    for i in range(0, total, 10000):
+        f = open('pubmed_batch_'+str(i)+'_to_'+str(i+9999)+".json", 'r')
+        data_file = f.read()
+        p_flag = False
+        for word in data_file.split():
+            if word == 'pmid':
+                p_flag = True
+            elif p_flag == True:
+                p_flag = False
+                if word[:-1] not in store:
+                    store.append(word[:-1])
+        f.close()
+    return store
+
+def main():
+    ab = open("test.json", 'w')
+    store = get_list_pmid("2021/01/17","2021/01/17")
+    print(store)
+    print(len(store))
+    here = uids_to_docs(store)
+    for item in here:
+        ab.write(json.dumps(item))
+        ab.write("\n")
+    ab.close()
+
+if __name__ == "__main__":
+    main()
```
Is most of this new logic beyond what @jvwong has written?

JohnGiorgi commented:
@Anwesh1 We should add the [.env](https://github.com/PathwayCommons/semantic-search/blob/master/.env) file from semantic search (minus whatever constants we don't need).

JohnGiorgi commented on pubmed-semantic(updated).py:
```diff
@@ -0,0 +1,286 @@
+        f.close()
+    return store
+
+def main():
+    ab = open("test.json", 'w')
+    store = get_list_pmid("2021/01/17","2021/01/17")
+    print(store)
+    print(len(store))
+    here = uids_to_docs(store)
+    for item in here:
+        ab.write(json.dumps(item))
+        ab.write("\n")
+    ab.close()
+
+if __name__ == "__main__":
+    main()
```
Okay so in general we should keep the code and the tests in different files (we can talk about setting up unit tests sometime if you like!).

I would remove this. If you like, you can create a docstring for `get_list_pmid` and put a simple hardcoded example in there.

Anwesh1 replied:
Okay I see, I can use another file to test it. Sure, we can talk about unit tests when you are free, I appreciate that.

JohnGiorgi replied:
The short version: I have a strong preference for [`pytest`](https://docs.pytest.org/en/stable/). It is very easy to pick up and get started with.

jvwong commented:
> The pubmed-semantic(updated).py is another file that's essentially taking in a list of around 2635 PMID's however there's an error thrown by the uid_to_docs() that's apparently a validation error. This is a little weird because when I use a subset of about 10 entries then it works fine, but I think one of the PMID's doesn't have a title and that's causing an error.

Every Article is supposed to have a ArticleTitle: see https://dtd.nlm.nih.gov/ncbi/pubmed/doc/out/180101/el-Article.html 

Can you attach your `test.json` file or better yet, find the offending UID?

Anwesh1 commented:
> > The pubmed-semantic(updated).py is another file that's essentially taking in a list of around 2635 PMID's however there's an error thrown by the uid_to_docs() that's apparently a validation error. This is a little weird because when I use a subset of about 10 entries then it works fine, but I think one of the PMID's doesn't have a title and that's causing an error.
> 
> Every Article is supposed to have a ArticleTitle: see https://dtd.nlm.nih.gov/ncbi/pubmed/doc/out/180101/el-Article.html
> 
> Can you attach your `test.json` file or better yet, find the offending UID?

Oh okay. The `test.json` file isn't having anything written to it because of the error. Unfortunately, it is hard for me to pinpoint the  UID that is causing the problem as I'm just passing the whole list. I'm trying to still pinpoint it by changing the number of records passed.

Anwesh1 commented:
I think I pinpointed it to the UID -> 33452616

I also think the problem is because the PMID doesn't have a proper title. Under the title tag, you can see that there's only a 'trans'.

![image](https://user-images.githubusercontent.com/57552053/105441930-918fa900-5c37-11eb-9720-4210c3092024.png)
 
However a normal record usually has a name just like this.

![image](https://user-images.githubusercontent.com/57552053/105442510-da476200-5c37-11eb-8d03-6af67e49f5cd.png)

The `uids_to_docs()` isn't able to find an ArticleTitle for it I think. It's a tricky one.

jvwong commented:
> I think I pinpointed it to the UID -> 33452616
> 
> I also think the problem is because the PMID doesn't have a proper title. Under the title tag, you can see that there's only a 'trans'.
> 
> ![image](https://user-images.githubusercontent.com/57552053/105441930-918fa900-5c37-11eb-9720-4210c3092024.png)
> 
> However a normal record usually has a name just like this.
> 
> ![image](https://user-images.githubusercontent.com/57552053/105442510-da476200-5c37-11eb-8d03-6af67e49f5cd.png)
> 
> The `uids_to_docs()` isn't able to find an ArticleTitle for it I think. It's a tricky one.

That's a weird entry. It has a self-closing tag `<ArticleTitle/>` -  I guess that's a edge case for us to consider ( but in my opinion NCBI is breaking its own schema DTD). 

Maybe some code to check if the title/abstract exists and skip if not at least a title?

Anwesh1 commented:
@jvwong, I'm not exactly sure how your whole script works, is there a specific place where I can set a flag that only returns the article if there is an article title? Also, I had made all the code for checking for Abstract and title and only grabbing the ID if both were present? Do you want me to make this again on your script?
It would help a lot if you could provide some insight into your script. I tried changing some of the `_articles_to_docs()` but it didn't seem to work.

jvwong committed:
Add .env file with basic defaults

jvwong commented:
> @jvwong, I'm not exactly sure how your whole script works, is there a specific place where I can set a flag that only returns the article if there is an article title? Also, I had made all the code for checking for Abstract and title and only grabbing the ID if both were present? Do you want me to make this again on your script?
> It would help a lot if you could provide some insight into your script. I tried changing some of the `_articles_to_docs()` but it didn't seem to work.

Hi @Anwesh1 two things:

1. I added an .env file. The reason it stopped you prior to this is because the `.gitignore` file, which instructs git to disregard  specified files, had an entry for anything named `.env`. When developing locally, you can always temporarily edit the key in this file  but don't check it into github (so the world can't steal your key :) )

2. Your troublesome PubMed id "33452616". 

The short of it is to keep the code as is - it is indeed correctly validating. The solution is to only provide it only with uids for articles in english using a [search field url parameter](https://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.ESearch) :

Sample ESEARCH call
```
https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?retmode=json&db=pubmed&datetype=pdat&reldate=1&term=(eng[Language])
```

This last call filtered out 20 articles.

jvwong commented:
This issue was helpful also because it raises the issue of how you're going to monitor the progress of the script and how to figure out which ones you did or didn't retrieve.

These are just a couple of the tradeoffs you encounter when you 1. Want to retrieve things in batch and 2. Use the live EUTILS rather than downloading the daily update files.

Anwesh1 commented:
Hey @jvwong, I see. This seems like it should work because I can see that it has filtered out some articles. However, it still keeps the article that was causing us a problem (33452616) even though that doesn't look like English. For some reason the filter isn't filtering that specific UID out, which is weird.

Anwesh1 commented:
Hey @jvwong, that was a perfect fix!! thanks for that.

So as a test, I ran your script for data collected on 1 day (19th Jan 2021), which had about 5597 records. It took 64.49 seconds to run. This is a snip of the output
`{"uid": "number", "title": "text", "abstract": text}` 
I made it output the title and abstract separate but we can revert back to the original if you'd like.

Then I ran my script on the same day for the same amount of records. It took 22.224 seconds to run. This is a snip of the output
`PMID \tab Title \tab Abstract`
My script is not grabbing data if the record doesn't have an abstract but we can change that as per the requirement.

jvwong commented:
Yeah my script fetches XML and is capable of finding a lot of detailed information (e.g. author forename, lastname, initials) so its bound to be slower. 

One thing you could try to do is add a search filter to include only those with [publication type "Journal Article"](https://www.nlm.nih.gov/mesh/pubtypes.html). 

https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?retmode=json&db=pubmed&datetype=pdat&reldate=1&term=(eng[Language])+AND+(Journal+Article[Publication+Type])

Anwesh1 commented:
@jvwong Is that the only type we want? Ones that have "Journal Article" as their publication type? 
It works perfectly with that URL thanks again.

Anwesh1 commented:
> Yeah my script fetches XML and is capable of finding a lot of detailed information (e.g. author forename, lastname, initials) so its bound to be slower.
> 
> One thing you could try to do is add a search filter to include only those with [publication type "Journal Article"](https://www.nlm.nih.gov/mesh/pubtypes.html).
> 
> [https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?retmode=json&db=pubmed&datetype=pdat&reldate=1&term=(eng[Language])+AND+(Journal+Article[Publication+Type])](https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?retmode=json&db=pubmed&datetype=pdat&reldate=1&term=(eng%5BLanguage%5D)+AND+(Journal+Article%5BPublication+Type%5D))

Hey, @jvwong so do we want our information only from ones that have "Journal Article" as their publication type? Also, I was wondering if we want the title and abstract together or as separate tags? Now the only thing left is to run the whole thing I'm assuming?

jvwong commented:
> Hey, @jvwong so do we want our information only from ones that have "Journal Article" as their publication type? 

Makes sense to me.

> Also, I was wondering if we want the title and abstract together or as separate tags? 

Together is fine for me, since we're interested in comparing article text to article text.

> Now the only thing left is to run the whole thing I'm assuming?

My suggestion is to update the PR #4:

- Create a README.md file. Fill it in with necessary and sufficient instructions in order that some new developer would be able to run your script and understand its capabilities (see semantic-search)
- Get a version that you would want to merge into the main branch
  - Clean up the files (there's two named `pubmed-semantic*`
  - Drop the `main` functions 
  - etc

Anwesh1 commented:
Hey, @jvwong Okay that sounds good. I was trying to run the script for the data produced in a week and there are some validation errors sent out by your script. I feel like whenever there's an empty title or abstract the script seems to break (we fixed a few of those when we only started dealing with stuff in English, but not all). This is a snip of the error:

![image](https://user-images.githubusercontent.com/57552053/105768828-f2b2c780-5f2a-11eb-99d7-6d84552ad456.png)

 Also, the XML is taking a considerably longer time to run. Any suggestions possibly?

jvwong commented:
> Hey, @jvwong Okay that sounds good. I was trying to run the script for the data produced in a week and there are some validation errors sent out by your script. I feel like whenever there's an empty title or abstract the script seems to break (we fixed a few of those when we only started dealing with stuff in English, but not all). This is a snip of the error:
> 
> ![image](https://user-images.githubusercontent.com/57552053/105768828-f2b2c780-5f2a-11eb-99d7-6d84552ad456.png)
> 
> Also, the XML is taking a considerably longer time to run. Any suggestions possibly?

It helps if you can identify the cases. Right now, the script is validating if there is a an `ArticleTitle`. It allows for empty Abstract, as specified by their [DTD ](https://dtd.nlm.nih.gov/ncbi/pubmed/out/pubmed_190101.dtd). I'd be curious to see what those cases look like.

Anwesh1 commented:
Hey @jvwong, I went through all the records and I believe there are 2 cases that are causing problems. 

The first one is the PMID -> 33279447
This one I believe is because it classifies its language as English and Spanish and doesn't have a title, like before. So we might need to some how set up that only English language is allowed and no combined cases (not yet sure how though).

![image](https://user-images.githubusercontent.com/57552053/105774546-5ccf6a80-5f33-11eb-8b12-3d2de0b9847b.png)

The second one is the PMID -> 33278872
This one is tricky for some reason. It has all its data like normal, however, there is an error. These are the errors for this PMID

![image](https://user-images.githubusercontent.com/57552053/105774763-b59f0300-5f33-11eb-98bc-c39092305c3e.png)

jvwong committed:
- Skip over AbstractText elements that are empty
- Allow empty title

jvwong commented:
> Hey @jvwong, I went through all the records and I believe there are 2 cases that are causing problems.
> 
> The first one is the PMID -> 33279447
> This one I believe is because it classifies its language as English and Spanish and doesn't have a title, like before. So we might need to some how set up that only English language is allowed and no combined cases (not yet sure how though).
> 
> The second one is the PMID -> 33278872
> This one is tricky for some reason. It has all its data like normal, however, there is an error. These are the errors for this PMID
> 

This was useful stuff for updating this and the semantic-search module (https://github.com/PathwayCommons/semantic-search/issues/46) thanks! The first PMID had a self-closing `ArticleTitle` and the second a self-closing `AbstractText`. I still believe that the first one breaks their spec and the second is useless but anyways I made the code accept these cases.

See aeeb125579f3847f42c8d65c1405435996335e56

jvwong commented on pubmed-semantic.py:
```diff
@@ -0,0 +1,250 @@
+                raise RuntimeError(ERROR)
+            pubmedEfetchReponse = PubmedEfetchResponse(**eutil_response)
+            pubmedArticle = pubmedEfetchReponse.PubmedArticleSet.PubmedArticle
+        except Exception as e:
```
Just FYI: When something fails validation, it throws a [`ValidationError` ](https://pydantic-docs.helpmanual.io/usage/validators/) that you can handle. Currently, it is not being handled - just kills the process, but you could handle it here (e.g. track the ids you missed) and continue the loop.

Anwesh1 commented:
Hey @jvwong, thanks for the changes, I updated them accordingly and they fix the ones we were having issues with. However, there's a new bug on the PMID -> 33303698. and this is a snip of the bug. I think it is running into a problem with one of the changes that you made to fix the previous issue.

![image](https://user-images.githubusercontent.com/57552053/105908833-b0ee5380-5ff4-11eb-9cc5-200a3c208252.png)

Do you know why this might be happening? The data looks fine and is in the format we want (title and abstract present and in English).

jvwong committed:
- Retrieve files using the medline rettype
- use the Biopython package Medline function to parse

jvwong committed:
have the eutil response return a generator

jvwong commented:
> Hey @jvwong, thanks for the changes, I updated them accordingly and they fix the ones we were having issues with. However, there's a new bug on the PMID -> 33303698. and this is a snip of the bug. I think it is running into a problem with one of the changes that you made to fix the previous issue.
> 
> Do you know why this might be happening? The data looks fine and is in the format we want (title and abstract present and in English).

Long story short the problem was that the xml library was mangling all the text. Good catch, this also affects the semantic-search code. 

I cleaned up the implementation and made it much simpler. Basically it grabs a ['medline'](https://www.nlm.nih.gov/bsd/mms/medlineelements.html) version of each article record and let [Biopython Medline](https://biopython.org/docs/1.75/api/Bio.Medline.html) do all the parsing. The actual data picking happens in `_medline_to_docs`.

I also had to update to get the `.env` file working.

Look around, make sure you think it makes sense then run some tests. I'd be curious how fast it is.

Anwesh1 commented:
Hey @jvwong, thanks for the fix. This makes it much simpler, no doubt. It is running perfectly now. I went through it and tested it for a month's worth of data and it works great. However, I think the time aspect of it is still a little slow, it took 1745 seconds for about 110k entries. Otherwise, it's good.

I also changed the .env file to what you had updated it to, but it caused a Bad request error. So, I just kept it the same as before and it works perfectly.

I wrote the README.md file, it's not too much so I just put a screenshot here, let me know if you want me to add anything else to it.

![image](https://user-images.githubusercontent.com/57552053/106027153-9cfc2d80-6098-11eb-865e-088018a1df95.png)

You had also told me before to drop the main functions, by drop you meant to delete them right? I just need to add the function to get the list of PMID's to your script and it should be good to go?

jvwong commented:
> Hey @jvwong, thanks for the fix. This makes it much simpler, no doubt. It is running perfectly now. I went through it and tested it for a month's worth of data and it works great. However, I think the time aspect of it is still a little slow, it took 1745 seconds for about 110k entries. Otherwise, it's good.
> 
> I also changed the .env file to what you had updated it to, but it caused a Bad request error. So, I just kept it the same as before and it works perfectly.
> 
> I wrote the README.md file, it's not too much so I just put a screenshot here, let me know if you want me to add anything else to it.
> You had also told me before to drop the main functions, by drop you meant to delete them right? I just need to add the function to get the list of PMID's to your script and it should be good to go?


Can you push your changes to the remote so we can all see them?

Anwesh1 commented:
Hey @jvwong do you want me to push to this same branch? (semantic_search(ncbi-file))

jvwong commented:
> Hey @jvwong do you want me to push to this same branch? (semantic_search(ncbi-file))

Yeah, a branch/PR is like a work in progress, so feel free to push regularly. 

Also, it wasn't clear in the README how you actually get the ids, input them, run the system etc.

Anwesh1 committed:
Updated files

ncbi.py -> has all the methods, added the method to grab PMID's to the bottom of the file
pubmed-semantic.py -> has the main just for testing, will be removed before the merge
README.md -> completed mostly, please share/edit as you wish.

Anwesh1 committed:
Update .env

Anwesh1 committed:
Delete pubmed-semantic(updated).py

Anwesh1 commented:
I have made the changes, please let me know if any fixes are required. Can you add some in-depth details for the uids_to_docs in the README.md? Thanks

jvwong commented on .env:
```diff
@@ -0,0 +1,18 @@
+CUDA_DEVICE=1
+
+# Services
+NCBI_EUTILS_API_KEY=b99e10ebe0f90d815a7a99f18403aab08008
```
Remove this value, its top secret ;)

Anwesh1 replied:
Oh sorry, It's done.

Anwesh1 committed:
Update .env

Anwesh1 committed:
Update pubmed-semantic.py

Anwesh1 committed:
Rename pubmed-semantic.py to main.py

Anwesh1 committed:
Update README.md

Anwesh1 committed:
Update README.md

Anwesh1 committed:
Update README.md

JohnGiorgi commented on .env:
```diff
@@ -7,7 +7,7 @@ MAX_LENGTH=384
 CUDA_DEVICE=1
```
You can drop this. We don't need GPUs for fetching.

JohnGiorgi commented on .env:
```diff
@@ -0,0 +1,18 @@
+APP_VERSION=0.1.0
+APP_URL=https://semanticsearch.baderlab.org/
+ADMIN_EMAIL=info@biofactoid.org
+MAX_LENGTH=384
+CUDA_DEVICE=1
```
Drop

JohnGiorgi commented on .env:
```diff
@@ -0,0 +1,18 @@
+# General
+APP_NAME=semantic-search
```
Rename `APP_NAME`

JohnGiorgi commented on .env:
```diff
@@ -0,0 +1,18 @@
+# General
+APP_NAME=semantic-search
+APP_VERSION=0.1.0
+APP_URL=https://semanticsearch.baderlab.org/
```
Rename `APP_NAME` and `APP_URL`. Maybe to `pubmed-dl`. Not sure what `APP_URL` should be, maybe just set it to this repo's URL for now.

JohnGiorgi commented on .env:
```diff
@@ -0,0 +1,18 @@
+EUTILS_ESUMMARY_BASENAME = "esummary.fcgi"
+HTTP_REQUEST_TIMEOUT = 5
+
+# Docker - NO GPU support via compose (https://github.com/NVIDIA/nvidia-docker)
+SEMANTIC_SEARCH_IMAGE_TAG=latest
+SEMANTIC_PORT=8000
```
Drop

Anwesh1 committed:
Update README.md

Anwesh1 committed:
Update README.md

Anwesh1 committed:
Update README.md

Anwesh1 committed:
Update .env

jvwong commented on README.md:
````diff
@@ -1 +1,50 @@
+*get_list_pmid*
+
+Works by taking in two strings that consist of the start and end date respectively in the YYYY/MM/DD format (ex: "2020/01/10"). This method outputs the URL's for each batch of 10,000 records in the format of
+```
````
For mindate and maxdate, probably EFETCH defaults to publication date. There are other dates like created, etc...

jvwong reviewed (approved):
Awesome, works for me! I say lets merge it in.

Anwesh1 committed:
Update README.md

Anwesh1 commented:
So should I go ahead and merge with master??

jvwong commented:
> So should I go ahead and merge with master??

Sure. you can add other features in other PRs
